streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java

```diff
@@ -1363,6 +1365,16 @@ public void testWindowing() {
                 new KeyValueTimestamp<>(2, "L2+l2", 2002L),
                 new KeyValueTimestamp<>(3, "L3+l3", 2003L)
             );
+
+            //push two items with timestamp at grace edge; this should produce one join item, M0 is 'too late'
+            final long currentStreamTime = 2104;
+            final long lowerBound = currentStreamTime - timeDifference.toMillis() - grace.toMillis();
+            inputTopic1.pipeInput(0, "M0", lowerBound - 1);
+            inputTopic1.pipeInput(1, "M1", lowerBound + 1);
```

Comment:
Why `+ 1` -- `lowerBound` by itself should already be inclusive and produce a join result?

Reply:
I can see now that the naming is misleading.

The lowerbound is with regards to the grace period, 1900
However the lowerbound of the winow `1:l1` is 1901

So the +1 was there to make sure it is still within the window.

In general I start to wonder whether it wouldn't make more sense to test these two concerns (grace & windowing) separatley. E.g. with grace 150, `M0` is just a test case to assert that late records get dropped and `M1` is just another window bound test. With grace 104 we get the 'grace bound' and the 'l0 lower window bound' to overlap but it might be confusing. In other words, as you said 'this test aims to test window bounds'. So maybe I should move grace period tests into a separate test class?

Reply:
> The lowerbound is with regards to the grace period, 1900
> However the lowerbound of the winow 1:l1 is 1901

> So the +1 was there to make sure it is still within the window.

For this case, should `M0` get not have the `-1` to get a strict bound test? Comparing the other test code you updated, you also use `lowerBound -1` to a "late" record and `lowerBound` for still within grace period record. -- Thus, should `currentStreamTime = 2105` to make it all work as expected? -- There is some disconnect between this test setup and other test setup.

Reply:
> E.g. with grace 150, M0 is just a test case to assert that late records get dropped and M1 is just another window bound test. With grace 104 we get the 'grace bound' and the 'l0 lower window bound' to overlap but it might be confusing. In other words, as you said 'this test aims to test window bounds'. So maybe I should move grace period tests into a separate test class?

Well, yes and no. I general, I agree that we should "separate" window bound and grace-period test, but I believe we actually do this -- if we use grace 150, we don't worry about "late" data that might be dropped in the first phase of the test. And we bolt-on a last test case for grace (I don't think we would gain much to separate this into a single test method). -- The confusing part might be, that we cannot clear "only test grace period" w/o also (implicitly) test window bound... The grace test shows us, that a "late" record is dropped, and that a "still within grace bounds record is not dropped and this "still in grace bounds record" naturally falls at the window-start time of the window it belong into...